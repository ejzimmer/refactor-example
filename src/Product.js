import "./App.css"
import { ProductTypeSelector } from "./ProductTypeSelector"
import { fetchPostage } from "./fetchPostage"
import { useState } from "react"
import styled from "styled-components"
import { PRICES } from "./prices"

export function Product({ product }) {
  const [productType, setProductType] = useState("seed_packets")
  const [count, setCount] = useState(0)
  const [postage, setPostage] = useState(-1)

  const handleProductTypeChange = (productType) => {
    setProductType(productType)
  }

  const handleCountChange = (event) => {
    setCount(event.currentTarget.value)
    calculatePostage(event.currentTarget.value, setPostage)
  }

  const calculatePostage = async (count, setPostage) => {
    if (productType == "seed_packets") return

    if (count == 0) {
      setPostage(0)
    }

    const postage = await fetchPostage(count)
    setPostage(postage)
  }

  return (
    <Container>
      <Heading>{product}</Heading>
      <ProductImage src={`${product}.jpg`} alt={product} />
      <Form>
        <ProductTypeSelector
          onChange={handleProductTypeChange}
          value={productType}
        />
        <div>
          {productType == "seed_packets" && (
            <label htmlFor="count">
              Number of seed packets (${PRICES.SEED_PACKETS} * {count} = $
              {PRICES.SEEDS * count})
            </label>
          )}
          {productType == "seedlings" && (
            <label htmlFor="count">
              Number of seedlings (${PRICES.SEEDLINGS} * {count} = $
              {PRICES.SEEDLINGS * count})
            </label>
          )}
          <br />
          <input id="count" type="number" onChange={handleCountChange} />
        </div>
        {productType == "seed_packets" && <Postage>Postage = $8</Postage>}
        {productType != "seed_packets" && postage >= 0 && (
          <Postage>Postage = ${postage}</Postage>
        )}
        {productType != "seed_packets" && postage < 0 && (
          <Error>Could not calculate postage. Please contact us.</Error>
        )}

        {productType == "seed_packets" && (
          <Total>Total: ${PRICES.SEED_PACKETS * count + 8}</Total>
        )}
        {productType != "seed_packets" && (
          <Total>Total: ${PRICES.SEEDLINGS * count + postage}</Total>
        )}
      </Form>
    </Container>
  )
}

const Container = styled.section`
  display: grid;
  grid-template-columns: 30vw 30vw;
  gap: 20px;
`

const Heading = styled.h2`
  grid-column: 1/-1;
  text-transform: capitalize;
`

const ProductImage = styled.img`
  width: 100%;
`

const Form = styled.form`
  display: flex;
  flex-direction: column;
  gap: 20px;
`

const Postage = styled.div`
  color: green;
`
const Error = styled.div`
  color: red;
`
const Total = styled.div`
  font-weight: bold;
`
