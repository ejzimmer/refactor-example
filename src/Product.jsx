import "./App.css"
import { ProductTypeSelector } from "./ProductTypeSelector"
import { fetchPostage } from "./fetchPostage"
import { useState, useEffect } from "react"
import styled from "styled-components"
import { PRICES } from "./prices"

export function Product({ product }) {
  const [productType, setProductType] = useState("SEED_PACKETS")
  const [count, setCount] = useState("")
  const [postage, setPostage] = useState(0)

  const handleProductTypeChange = (productType) => {
    setProductType(productType)
  }

  const handleCountChange = (event) => {
    setCount(event.currentTarget.value)
  }

  useEffect(() => {
    calculatePostage()
  }, [productType, count])

  const calculatePostage = async () => {
    if (productType == "SEED_PACKETS") return

    if (!count) {
      setPostage(0)
    }

    const postage = await fetchPostage(count)
    setPostage(postage)
  }

  return (
    <Container>
      <Heading>{product}</Heading>
      <ProductImage src={`${product}.jpg`} alt={product} />
      <Form>
        <ProductTypeSelector
          onChange={handleProductTypeChange}
          value={productType}
        />
        <div>
          {productType == "SEED_PACKETS" && (
            <label htmlFor="count">
              Number of seed packets (${PRICES.SEED_PACKETS} * {count || 0} = $
              {PRICES.SEED_PACKETS * count || 0})
            </label>
          )}
          {productType == "SEEDLINGS" && (
            <label htmlFor="count">
              Number of seedlings (${PRICES.SEEDLINGS} * {count || 0} = $
              {PRICES.SEEDLINGS * count || 0})
            </label>
          )}
          <br />
          <input
            id="count"
            type="number"
            onChange={handleCountChange}
            value={count}
          />
        </div>
        {productType == "SEED_PACKETS" && <Postage>Postage = $8</Postage>}
        {productType != "SEED_PACKETS" && postage >= 0 ? (
          <Postage>Postage = ${postage}</Postage>
        ) : (
          productType != "SEED_PACKETS" && (
            <Error>Could not calculate postage. Please contact us.</Error>
          )
        )}

        {count === "" ? (
          <Total>Please enter a number to see the total</Total>
        ) : productType == "SEED_PACKETS" ? (
          <Total>Total: ${PRICES.SEED_PACKETS * count + 8}</Total>
        ) : (
          <Total>Total: ${PRICES.SEEDLINGS * count + postage}</Total>
        )}
      </Form>
    </Container>
  )
}

const Container = styled.section`
  display: grid;
  grid-template-columns: 30vw 30vw;
  gap: 20px;
`

const Heading = styled.h2`
  grid-column: 1/-1;
  text-transform: capitalize;
`

const ProductImage = styled.img`
  width: 100%;
`

const Form = styled.form`
  display: flex;
  flex-direction: column;
  gap: 20px;
`

const Postage = styled.div`
  color: green;
`
const Error = styled.div`
  color: red;
`
const Total = styled.div`
  font-weight: bold;
`
