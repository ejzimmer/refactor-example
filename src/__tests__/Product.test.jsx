import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Product } from '../Product'

describe('Product component', () => {
  describe('when seed packets is selected with no amount', () => {
    it('matches the snapshot', () => {
      const { container } = render(<Product product="corn" />)

      const radioButton = screen.getByRole('radio', { name: 'Seed Packets' })
      userEvent.click(radioButton)

      expect(container).toMatchSnapshot()
    })
  })

  describe('when seed packets is selected with an amount', () => {
    it('matches the snapshot', () => {
      const { container } = render(<Product product="corn" />)

      const radioButton = screen.getByRole('radio', { name: 'Seed Packets' })
      userEvent.click(radioButton)

      const input = screen.getByRole('spinbutton')
      userEvent.type(input, '4')

      expect(container).toMatchSnapshot()
    })
  })

  describe('when seedlings is selected with no amount', () => {
    it('matches the snapshot', () => {
      const { container } = render(<Product product="corn" />)

      const radioButton = screen.getByRole('radio', { name: 'Seedlings' })
      userEvent.click(radioButton)

      expect(container).toMatchSnapshot()
    })
  })

  describe('when seedlings is selected with an amount', () => {
    it('matches the snapshot', () => {
      const { container } = render(<Product product="corn" />)

      const radioButton = screen.getByRole('radio', { name: 'Seedlings' })
      userEvent.click(radioButton)

      const input = screen.getByRole('spinbutton')
      userEvent.type(input, '4')

      expect(container).toMatchSnapshot()
    })
  })

})