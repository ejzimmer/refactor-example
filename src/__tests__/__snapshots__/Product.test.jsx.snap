// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product matches the snapshot when seed packets is selected and a number is entered 1`] = `
<div>
  <section
    class="sc-hKwDye bHYLkR"
  >
    <h2
      class="sc-eCImPb opaLQ"
    >
      corn
    </h2>
    <img
      alt="corn"
      class="sc-jRQBWg csmCrb"
      src="corn.jpg"
    />
    <form
      class="sc-gKclnd jlcuhb"
    >
      <fieldset
        style="display: flex; gap: 1.5em;"
      >
        <legend>
          Product type
        </legend>
        <label>
          <input
            checked=""
            name="product_type"
            type="radio"
            value="SEED_PACKETS"
          />
          Seed Packets
        </label>
        <label>
          <input
            name="product_type"
            type="radio"
            value="SEEDLINGS"
          />
          Seedlings
        </label>
      </fieldset>
      <div>
        <label
          for="count"
        >
          Number of 
          seed packets
           ($
          4
           * 
          4
           = $
          16
          )
        </label>
        <br />
        <input
          id="count"
          type="number"
          value=""
        />
      </div>
      <div
        class="sc-bdvvtL iGuzkS"
      >
        Postage = $
        8
      </div>
      <div
        class="sc-dkPtRN kRyTEO"
      >
        Total: $
        24
      </div>
    </form>
  </section>
</div>
`;

exports[`Product matches the snapshot when seed packets is selected and no number is entered 1`] = `
<div>
  <section
    class="sc-hKwDye bHYLkR"
  >
    <h2
      class="sc-eCImPb opaLQ"
    >
      corn
    </h2>
    <img
      alt="corn"
      class="sc-jRQBWg csmCrb"
      src="corn.jpg"
    />
    <form
      class="sc-gKclnd jlcuhb"
    >
      <fieldset
        style="display: flex; gap: 1.5em;"
      >
        <legend>
          Product type
        </legend>
        <label>
          <input
            checked=""
            name="product_type"
            type="radio"
            value="SEED_PACKETS"
          />
          Seed Packets
        </label>
        <label>
          <input
            name="product_type"
            type="radio"
            value="SEEDLINGS"
          />
          Seedlings
        </label>
      </fieldset>
      <div>
        <label
          for="count"
        >
          Number of 
          seed packets
           ($
          4
           * 
          0
           = $
          0
          )
        </label>
        <br />
        <input
          id="count"
          type="number"
          value=""
        />
      </div>
      <div
        class="sc-bdvvtL iGuzkS"
      >
        Postage = $
        8
      </div>
      <div
        class="sc-dkPtRN kRyTEO"
      >
        Please enter a number to see the total
      </div>
    </form>
  </section>
</div>
`;

exports[`Product matches the snapshot when seedlings is selected and a number is entered 1`] = `
<div>
  <section
    class="sc-hKwDye bHYLkR"
  >
    <h2
      class="sc-eCImPb opaLQ"
    >
      corn
    </h2>
    <img
      alt="corn"
      class="sc-jRQBWg csmCrb"
      src="corn.jpg"
    />
    <form
      class="sc-gKclnd jlcuhb"
    >
      <fieldset
        style="display: flex; gap: 1.5em;"
      >
        <legend>
          Product type
        </legend>
        <label>
          <input
            checked=""
            name="product_type"
            type="radio"
            value="SEED_PACKETS"
          />
          Seed Packets
        </label>
        <label>
          <input
            name="product_type"
            type="radio"
            value="SEEDLINGS"
          />
          Seedlings
        </label>
      </fieldset>
      <div>
        <label
          for="count"
        >
          Number of 
          seedlings
           ($
          8
           * 
          4
           = $
          32
          )
        </label>
        <br />
        <input
          id="count"
          type="number"
          value=""
        />
      </div>
      <div
        class="sc-bdvvtL iGuzkS"
      >
        Postage = $
        32
      </div>
      <div
        class="sc-dkPtRN kRyTEO"
      >
        Total: $
        64
      </div>
    </form>
  </section>
</div>
`;

exports[`Product matches the snapshot when seedlings is selected and no number is entered 1`] = `
<div>
  <section
    class="sc-hKwDye bHYLkR"
  >
    <h2
      class="sc-eCImPb opaLQ"
    >
      corn
    </h2>
    <img
      alt="corn"
      class="sc-jRQBWg csmCrb"
      src="corn.jpg"
    />
    <form
      class="sc-gKclnd jlcuhb"
    >
      <fieldset
        style="display: flex; gap: 1.5em;"
      >
        <legend>
          Product type
        </legend>
        <label>
          <input
            checked=""
            name="product_type"
            type="radio"
            value="SEED_PACKETS"
          />
          Seed Packets
        </label>
        <label>
          <input
            name="product_type"
            type="radio"
            value="SEEDLINGS"
          />
          Seedlings
        </label>
      </fieldset>
      <div>
        <label
          for="count"
        >
          Number of 
          seedlings
           ($
          8
           * 
          0
           = $
          0
          )
        </label>
        <br />
        <input
          id="count"
          type="number"
          value=""
        />
      </div>
      <div
        class="sc-bdvvtL iGuzkS"
      >
        Postage = $
        0
      </div>
      <div
        class="sc-dkPtRN kRyTEO"
      >
        Please enter a number to see the total
      </div>
    </form>
  </section>
</div>
`;
